apps/prairielearn/src/middlewares/authzCourseOrInstance.js

```diff
@@ -523,6 +535,35 @@ module.exports = asyncHandler(async (req, res, next) => {
       `.toString();
       throw err;
     }
+
+    // The effective user is not enrolled in the course instance and is also not
+    // either a course instructor or a course instance instructor - remove all
+    // override cookies and return with error.
+    //
+    // Note that we skip this check if the effective user is the same as the
+    // authenticated user, since an instructor may want to view their course
+    // as a student without enrolling in their own course.
+    if (
+      !idsEqual(user.uid, res.locals.authn_user.uid) &&
+      !effectiveResult.rows[0].permissions_course.has_course_permission_preview &&
+      !effectiveResult.rows[0].permissions_course_instance.has_course_instance_permission_view &&
+      !effectiveResult.rows[0].permissions_course_instance.has_student_access_with_enrollment
```

Comment:
Let's double-check this conditional. Specifically, I'm wondering if it's _too_ precise at the moment. That is, consider the case where an instructor is impersonating another user that is _not_ enrolled and also requesting "without access restrictions" - that is, with the authenticated user's role. I'd like to make sure that the instructor can't then take actions as the student. I should probably write a test for this...

Reply:
My understanding is that at this point we known that the authn user is course staff. There are then four important cases:
1. If effective user is the same as authn user --> allow access.
2. Otherwise, if effective user is also course staff --> allow access.
3. Otherwise, if effective user is an enrolled student --> allow access.
4. Otherwise block access.

I think the above if statement is correctly blocking access only in case 4?

Reply:
Thanks for checking!